### ISSUE TITLE
A jump drive with changeable range : Feature Request

### ISSUE THREAD

**Issue by Wrzlprnft** [endless-sky/endless-sky#788]
Even if i do not support the idea of a galaxy-wide-jump-drive-range (http://steamcommunity.com/app/404410/discussions/2/412446890563414360/?tscn=1455448887) i think a adjustable jump drive range would be a part of the aliens technology ladder, and would also open up space for modders.

I looked in the source code but figured out nothing, so if i've overseen something just tell me :)

Oherwise maybe intialise the jump drive as normal, but maybe add a range attribute, and make another if-clause in the method which calculates the jump.

I think the fuel cost should be directly connected to the range though.

**Comment by LocalGod79** [comment]
fuel:range ratio for a single jump should perhaps grow more rapidly than 'directly' - exponentially (to the power of 1.5, perhaps)?

this would ensure that galaxy wide jumps would be an extreme rarity (if even possible)...

**Comment by DingusShingleton** [comment]
Enter : Bactrian with 100 fuel cells

**Comment by jmhorjus** [comment]
I think the source code value you're looking for is the hard-coded constant "static const double NEIGHBOR_DISTANCE = 100.;" in System.cpp.  It determines which systems are considered "adjacent" for purposes of jump drives, but is currently universal - i.e. all jump drives and map code use the same constant distance and the "neighbors" of a system are calculated only once when the map is loaded. 

If you're able to build the source code, you could start by modifying NEIGHBOR_DISTANCE and observing the effects.  To make that distance variable during run-time you'd probably want to replace the system's "std::vector<const System *> neighbors;" (a vector of neighboring systems) with a vector of vectors or a map of vectors indexed by the neighbor_distance value used to calculate them.  This would work well if you had only a few different ranges of jump drive, which seem likely how you would want to have it.  
The map could possibly be populated based on the different ranges of drives found in the outfits data so it would still only happen once when the map was loaded. 

Then you'd have to find all the places where the neighbors list is used and update them.  Right now "system.neighbors" is already only used if you have a jump drive (otherwise "system.links" is used), so you should be able to pass in a parameter indicating your range so it knows what list of neighbors to use.  

Happy hacking!

**Comment by Wrzlprnft** [comment]
@jmhorjus thx man! i just was to fixed on the idea of a range-attribute somewhere

additionally: it was not my idea, the steam forum came up with this, but i considered it worth putting it here. i guess the way it should work would be maybe several rows of neighbour-classified systems? like 100-200-300 etc... or would that eat up the memory? resulting would be an additional attribute for the jump drive, classified as one two three or something like that

**Comment by jmhorjus** [comment]
Yep, that's the general idea. All the details you need to work out will come into focus once you start the process of building/testing/iterating. 
Have you been able to make a fork, go through the build instructions, (https://github.com/endless-sky/endless-sky/wiki/BuildInstructions) and successfully build ES from your own fork? 

Be advised that this is a somewhat complicated C++ programming task you're assigning yourself.  If that's not your specific skillset, you might consider contributing new missions/stories or art instead - content like missions, ships, outfits, planets, star systems, etc, that use existing attributes can all be made without changing the actual C++ source at all.

**Comment by thenerdfreak** [comment]
I think a way to fix this would be to just make those "further away" systems that require a "better jumpdrive" require that jumpdrive to access it. like a special "has*nameofjumpdrive*equipped" parameter or whatever it's called in called. Maybe a slightly higher jump speed to go along with it

**Comment by Wrzlprnft** [comment]
I keep getting errors with the gcc.exe, but ill fix them once i have the time.
I started studying IT last year october, so i guess i have to leran C++ at some point, so ill do it on a project i like more than boring uni stuff.
The real problem is i am a stranger to the source code of endless sky itself, and my limited C++ knowledge makes it hard to follow all the calls and references.... yet.
i guess i could copy parts of the already existing code, adjusting it for the different ranges, but how to organize the set up for the reading of the text-files and classifying of the different jump drives within the game.

**Comment by jmhorjus** [comment]
If you need build help post something on the forum here:
https://groups.google.com/forum/#!category-topic/endless-sky/bugs-and-troubleshooting/2tHEB8Nt3Ac
and I'll try to help if I can.

**Comment by Wrzlprnft** [comment]
Note: i won't finish this within this year i guess

**Comment by jmhorjus** [comment]
Sorry to hear it. :( Perhaps someone else will take up this idea and make it a reality.

**Comment by endless-sky** [comment]
Because of the way systems cache their jump drive "neighbors" to allow fast pathfinding, there's no way this could be done without significant code changes.

**Comment by Amazinite** [comment]
Code changes have been made.
